designs/2018-programmatic-style-edits/README.md

```diff
@@ -0,0 +1,362 @@
+
+Prettier has also shown that there is a desire to not fill up end-user screens with multiple error messages advising developers to add a space here or remove a space there; they'd much rather just have one command to fix the style of their source code.
+
+Additionally, a growing number of developers use ESLint together with Prettier to get both the error-checking capability of ESLint with the source code formatting of Prettier. While there are some approaches to making these two tools work together, any quick search of Twitter will find new developers tend to struggle to get these two tools to work together.
+
+All of this is to say that the current way that ESLint deals with style guides, including its interaction with Prettier, leaves a lot of room for improvement. To that end, this programmatic style editor feature is intended to:
+
+1. Allow arbitrary modification of source code within the context of autofixing.
+1. Enable the use of third-party tools to do this modification easily.
+1. Reduce developers' dependence on ESLint core rules for style guide enforcement and therefore reduce the maintenance burden on the ESLint team.
+
+## Detailed Design
+
+At a high-level, this design entails:
+
+1. Enable plugins to export style editors
+1. Allow third-party tools to be incorporated into style editors
+1. Automatically apply style editors with using `--fix`
+1. Allow turning on/off style editors using `--fix-type`
+1. Ensure style editors are composable
+
+Definitions:
+
+* **Style Edit** - an individual code modification written in JavaScript.
+* **Style Editor** - a collection of style edits to apply in order. Style editors are applied through configuration.
+
+### Style Editors in Plugins
+
+A plugin can contain one or more style editors, exported on a `styles` object in a manner similar to `rules` and `processors`, for example:
+
+```js
+// eslint-plugin-mystyles
+const es5StyleEditor = require("./styles/es5");
+
+module.exports = {
+    styles: {
+        es5: es5StyleEditor
+    }
+};
+```
+
+In this code, the plugin `eslint-plugin-mystyles` exports a style editor called `es5`. In order to apply the style editor to code, the end-user would add a `style` property to the configuration object:
+
+```js
+module.exports = {
+    plugins: ["mystyles"],
+    parserOptions: {
+        ecmaVersion: 5
+    },
+    style: "mystyles/es5"
+}
+```
+
+The `style` property can be used at the top level of the configuration to apply to all files or in an `overrides` object to apply to a subset of files.
+
+Options for style editors can be specified using a `styleOptions` key, as in this example:
+
+```js
+module.exports = {
+    plugins: ["mystyles"],
+    parserOptions: {
+        ecmaVersion: 5
+    },
+    style: "mystyles/es5",
+    styleOptions: {
+        indent: "tabs"
+    }
```

Comment:
I'm confused about the value-add here. Couldn't the same thing be accomplished by having the `mystyles` function export an `es5` *rule*, which modifies the entire contents of the file as an autofix?

(I realize that this would work a bit differently in terms of the order of operations, but the UX would largely be the same as far as I can tell.)

Reply:
Yes, and in fact, that's basically what a lot of our whitespace-related rules are doing right now.

The value-add, in my mind, is:

1. There's not a lot of value to printing out a warning about a style problem that can easily be fixed. I know I get really annoyed when I see a "missing space" or "incorrect indentation" warning. I'd rather it just be fixed and not bother me. 
1. The `fix` API is fairly limited in what it can do right now. As I've mentioned before, it was really designed for making tweaks here and there but we are now using it for all kinds of changes that it wasn't designed for (think function expressions to arrow function expressions). It's really difficult to change how rule fixes work because of the single traversal architecture, which is why we have to apply fixes after linting instead of during. Style edits are realtime changes that basically have no limitations.

Reply:
Thanks for explaining.

I'm not sure I understand point 2. To me, it seems like the current fix API is difficult to use not because it's limited in expressive power (rules can already replace the entire contents of the file if they want to), but rather because the API operates at a low level of abstraction (it only allows raw text-based replacements rather than AST-based replacements). In principle, rules could create autofixers that use `recast` for formatting, although I'm not aware of any rules that do this at the moment.

It seems to me that style editors would be similarly low-level in that they would simply provide a text-based API. So I don't see how this would be solving any problems with the current `fix` API that wouldn't already be solvable by having a rule that does something like this:

```js
context.report({
    /* ... */
    fix: fixer => fixer.replaceTextRange([0, sourceText.length], newSourceText)
});
```

Reply:
The goal here is not to workaround limitations of fixers, it's to create a new way to make changes that are always applied (literally a source code formatter). It is not possible right now to say, "I want to make these changes to my source code but don't bug me about the individual tweaks that need to be made." Plus, this allows us to outsource small tweaks to style rules to the end users who want them rather than needing to create more and more options for core rules.

It helps to think about a concrete example: think of the `indent` rule. Everyone has an opinion about the right way to indent things, and we've tried to keep up by adding options for each preference. What if instead the `indent` rule just did basic indentation without exceptions, and then anyone who wanted to, for example, outdent `case` statements could just apply an edit to do that? This would free up a lot of time for the team. We'd be able to say, "we aren't making more changes to this rule, if you need to tweak the output, create a style edit." 

I can tell from your questions that I haven't done a good job outlining this motivation in the RFC, so I'll update the motivation section.